lib/pbench/server/api/resources/endpoint_configure.py

```diff
@@ -0,0 +1,166 @@
+
+                This is dynamically generated by processing the Flask URI
+                rules; refer to api/__init__.py for the code which creates
+                those mappings, or test_endpoint_configure.py for code that
+                validates the current set (and must be updated when the API
+                set changes). We supplement the Flask API list with the
+                "results" API, which is currently just an Apache public_html
+                file mapping but is referenced by the dashboard code.
+
+        TODO: We need an internal mechanism to track the active versions of the
+        various Elasticsearch template documents. We're hardcoding them here and
+        in other APIs. We should consider persisting an equivalent of the
+        mapping table built in "indexer.py" for use across the server APIs.
+
+        TODO: We provide Elasticsearch index root names here, which the dashboard
+        code needs to perform the queries we've not yet replaced with server-side
+        implementations. The entire "indices" section can be removed once that is
+        resolved.
+        """
+        self.logger.debug("Received these headers: {!r}", request.headers)
+        origin = None
+        host_source = "configuration"
+        host_value = self.host
+        header = request.headers.get("Forwarded")
+        if header:
+            m = self.forward_pattern.search(header)
+            if m:
+                origin = m.group("host")
+                host_source = "Forwarded"
+                host_value = header
+        if not origin:
+            header = request.headers.get("X-Forwarded-Host")
+            if header:
+                m = self.x_forward_pattern.search(header)
+                if m:
+                    origin = m.group("host")
+                    host_source = "X-Forwarded-Host"
+                    host_value = header
+        if not origin:
+            origin = self.host
+        host = f"http://{origin}"
+        self.logger.info(
+            "Advertising endpoints relative to {}: {}", host_source, host_value
+        )
+
+        # We pre-load the APIs list with the "results" link, which isn't yet
+        # a Pbench server API. The dashboard static endpoint configuration
+        # included this, and it makes sense for consistency.
+        apis = {"results": urljoin(host, "/results")}
+
+        # Iterate through the Flask endpoints to add a description for each.
+        for rule in current_app.url_map.iter_rules():
+            url = str(rule.rule)
+
+            # Ignore anything that doesn't use our API prefix, because it's
+            # not in our API.
+            if url.startswith(self.uri_prefix):
+                # If the URI is parameterized with a Flask "<type:name>"
+                # template string, we don't want to report it, so we remove
+                # it from the URI. We derive an API name by converting the
```

Comment:
The client has to remove/substitute the templates then and I figured that wasn't very generic. I didn't really want to make it so Javascript-specific, either, and Javascript will only do that sort of substitution in literal template strings (backticked), so far as I can tell; although it's just a `.replace` away, it's not quite that neat.

Anyway, just letting it append the username to end end of the URI seemed easiest, at least for now.

Reply:
I second the idea of not making this Javascript specific.

Reply:
I guess I was proposing two things:

- make the URI available as a template
- make it easy for the client to fill in the template (which is why I offered JS syntax)

It doesn't have to be Javascript format, but I'll point out that that format also works for shell, so it's not exactly JS-specific.  ;-)  We could use `%s` instead of `${...}`, but I opted for the latter because it provides semantic information.

It's true that the client will likely need to have some foreknowledge in order to fill in the template, but it will need to have even closer coupling to produce the URI if we just lop it off at the first parameter.  Providing a template removes the need for some of that coupling.

Apparently there is a (proposed) IETF [standard for URI templates](https://tools.ietf.org/html/rfc6570).

Reply:
Let's track in issue #2154.